Cli/Test/EnvironmentTests.cs

```diff
@@ -0,0 +1,97 @@
+
+        public EnvironmentTests(ITestOutputHelper testOutputHelper)
+        {
+            _output = testOutputHelper;
+        }
+
+        [Fact]
+        public void VerifyDuplicateEnvironmentVariablesDoNotCrashGitCredentialManager()
+        {
+            ServicePointManager.SecurityProtocol = ServicePointManager.SecurityProtocol | SecurityProtocolType.Tls12;
+
+            var workingDir = new FileInfo(Assembly.GetExecutingAssembly().Location).Directory.FullName;
+            var exePath = Path.Combine(workingDir, "git-credential-manager");
+            var startInfo = GetProcessStartInfoWithCaseInsensitiveEnvVars(exePath, "GET");
+            startInfo.EnvironmentVariables["casedvariablename"] = "B";
+            startInfo.EnvironmentVariables["CASEDVARIABLENAME"] = "b";
+
+            startInfo.EnvironmentVariables["GCM_MODAL_PROMPT"] = "false";
+            startInfo.EnvironmentVariables["GCM_INTERACTIVE"] = "false";
+
+            var process = new Process { StartInfo = startInfo };
+            process.StartInfo.RedirectStandardInput = true;
+            process.StartInfo.RedirectStandardOutput = true;
+            process.StartInfo.RedirectStandardError = true;
+            process.OutputDataReceived += CaptureOutput;
+            process.ErrorDataReceived += CaptureError;
+            if (process.Start())
+            {
+                process.BeginOutputReadLine();
+                process.BeginErrorReadLine();
+
+                using (var writer = process.StandardInput)
+                {
+                    writer.WriteLine("host=example.com");
+                    writer.WriteLine("protocol=https");
+                    writer.WriteLine();
+                    writer.WriteLine();
+                }
+
+                process.WaitForExit(1000);
+
+            }
+            
+            Assert.False(_envError);
+            
+        }
+
+        private void CaptureOutput(object sender, DataReceivedEventArgs e)
+        {
+            ShowOutput(e.Data, "output");
+        }
+
+        private void CaptureError(object sender, DataReceivedEventArgs e)
+        {
+            ShowOutput(e.Data, "ERROR");
+        }
+
+        private void ShowOutput(string data, string prefix)
+        {
+            if (data != null)
+            {
+                _output.WriteLine($"Received {prefix}: {data}");
+                
+                if(data.Contains("An item with the same key has already been added."))
+                {
+                    _envError = true;
+                }
+            }
+        }
+
+        private static ProcessStartInfo GetProcessStartInfoWithCaseInsensitiveEnvVars(string exePath, string args)
+        {
+            var startInfo = new ProcessStartInfo(exePath, args)
+            {
+                UseShellExecute = false,
+                CreateNoWindow = true
+            };
+            startInfo.SetFieldValue("environmentVariables", new CaseSensitiveStringDictionary());
```

Comment:
Reflection is almost never the right solution to anything, not even in a test 😄

Could we not just test that the `Settings::GetEnvironmentVariables(EnvironmentVariableTarget target)` method does not throw an exception when handed duplicate keys (differing by case only)?

Reply:
This would be my vote too.  It should be a very simple more specific test.

Reply:
I agree with your assessment that this would likely be the only way to get the duplicate keys in the environment block of the git-credential-manager.exe process. But I was thinking that we probably don't need an end-to-end test here.

Instead just a unit test around our `Settings::GetEnvironmentVariables` method ought to be enough coverage to ensure we don't regress (assuming that there aren't any parts of GCM that directly call `System.Environment::GetEnvironmentVariables`).

Reply:
I understand, but since the call the Environment is inside that method, how do we do that?

        public IDictionary<string, string> GetEnvironmentVariables(EnvironmentVariableTarget target)
        {
            var variables = Environment.GetEnvironmentVariables(target);

Apologies if I'm missing something obvious.

Reply:
Ah, no! My apologies – I'm the one who's missing things here :smile:

Short of using some P/Invoke to call the Win32 API to self-modify the env block in the tests (which is also pretty iffy)... I guess we could pull the `foreach` loop out into it's own helper method:

```csharp
IDictionary<string, string> DeduplicateStringDictionary(IDictionary variables)
```

..update `Settings::GetEnvironmentVariables` to be:

```csharp
public IDictionary<string, string> GetEnvironmentVariables(EnvironmentVariableTarget target)
{
    var variables = Environment.GetEnvironmentVariables(target);
    return DeduplicateStringDictionary(variables);
}
```

..and just test the dedup method standalone.